source/extensions/filters/http/cache/http_cache.cc

```diff
@@ -74,30 +73,67 @@ LookupRequest::LookupRequest(const Http::RequestHeaderMap& request_headers, Syst
 size_t stableHashKey(const Key& key) { return MessageUtil::hash(key); }
 size_t localHashKey(const Key& key) { return stableHashKey(key); }
 
-// Returns true if response_headers is fresh.
-bool LookupRequest::isFresh(const Http::ResponseHeaderMap& response_headers) const {
-  if (!response_headers.Date()) {
-    return false;
+bool LookupRequest::requiresValidation(const Http::ResponseHeaderMap& response_headers) const {
+  // TODO(yosrym93): Store parsed response cache-control in cache instead of parsing it on every
+  // lookup
+  absl::string_view cache_control =
+      response_headers.getInlineValue(response_cache_control_handle.handle());
+  const ResponseCacheControl response_cache_control(cache_control);
+
+  SystemTime response_time = CacheHeadersUtils::httpTime(response_headers.Date());
+
+  if (timestamp_ < response_time) {
+    // Response time is in the future, validate response
+    return true;
   }
-  const Http::HeaderEntry* cache_control_header =
-      response_headers.getInline(response_cache_control_handle.handle());
-  if (cache_control_header) {
-    const SystemTime::duration effective_max_age =
-        HttpCacheUtils::effectiveMaxAge(cache_control_header->value().getStringView());
-    return timestamp_ - HttpCacheUtils::httpTime(response_headers.Date()) < effective_max_age;
+
+  const SystemTime::duration response_age = timestamp_ - response_time;
+  bool request_max_age_exceeded = request_cache_control_.max_age_.has_value() &&
```

Comment:
Nit: more const everywhere in this PR. There are a lot of local variables that can be made immutable, please update them across the files of the PR. Thanks.